### ISSUE TITLE
[macOS] Cannot install some .air applications with shared runtime

### ISSUE THREAD

**Issue by itlancer** [airsdk/Adobe-Runtime-Support#990]
## Problem Description
Cannot install some .air applications with macOS AIR shared runtime.

I didn't understood exatly steps how to reproduce it but with earlier AIR SDK and AIR Shared Runtime 33.1 versions seems all worked fine for our users. But few weeks ago when we start to use AIR SDK 33.1.1.533/AIR SDK 33.1.1.554 and the same AIR Shared Runtimes our .air applications cannot be installed/updated anymore. At installation process we get error.

Meanwhile some other .air applications with the same signing certificate but different application IDs can installed and worked correctly. May be some applications IDs "history" remains. "Old versions" of .air applications (built with AIR SDK 33.1 but before 33.1.1.533) also cannot be installed anymore.
Problem with installation and updating (when previous version installed) .air application.

Tested with .air applications build with AIR SDK 33.1.1.533 and AIR SDK 33.1.1.554 (and some older one) with latest AIR Runtime 33.1.1.554.
.air applications uses AIR license to be built.

Uninstall and install AIR Runtime again didn't solve this issue.

Related issue:
https://github.com/airsdk/Adobe-Runtime-Support/issues/833
https://github.com/airsdk/Adobe-Runtime-Support/issues/565
https://github.com/airsdk/Adobe-Runtime-Support/issues/531

## Steps to Reproduce
Install AIR Shared Runtime 33.1.1.554 for macOS. Compile .air application and try to install it.
I can send application sample but not 100% users affected by this issue.

**Actual Result:**
![image](https://user-images.githubusercontent.com/10899066/125930395-e895879e-d4bf-40b3-b10b-44b5a62dd9e8.png)
"Sorry, an error has occurred." at the installer.
With OS Console you can see:
```
Jul 15 13:13:13 MacBook-Pro Console[832]: BUG in libdispatch client: vnode, monitored resource vanished before the source cancel handler was invoked { 0x600003308800[source], ident: 9 / 0x9, handler: 0x6b290001c7f9fc84 }
Jul 15 13:15:32 MacBook-Pro com.apple.xpc.launchd[1] (application.com.adobe.air.ApplicationInstaller.4395398.4395404[840]): Service exited with abnormal code: 7
Jul 15 13:15:36 MacBook-Pro com.apple.xpc.launchd[1] (com.apple.mdworker.shared.01000000-0500-0000-0000-000000000000[809]): Service exited due to SIGKILL | sent by mds[100]
Jul 15 13:15:37 MacBook-Pro com.apple.xpc.launchd[1]: Coalition Cache Hit: app<application.com.adobe.air.ApplicationInstaller.4395398.4395404(501)> [732]
```

**Expected Result:**
Application will be installed correctly without errors.

## Known Workarounds
none
Use native packaging (AIR bundle).

**Comment by denisgl7** [comment]
Same problem

**Comment by ajwfrost** [comment]
Hi - looking into this at the moment but not currently reproducing it. Are you able to confirm what macOS version(s) this is happening on? and if you can send us a .air file that's causing problems, we can try to reproduce it here.

From what we can see in code, there should have been some more log output (to syslog) prior to a failure with this error code, are you able to check via `syslog | grep "adobe\.air"` or does that just show the same output?

And do you have a completely reproducible set-up there so that e.g. we can provide a file with additional debug capabilities?

thanks

**Comment by itlancer** [comment]
@ajwfrost 

It reproduced with macOS 11.4 and macOS 11.3.1 different macbooks (Intel-based and Apple Silicon M1-based).

.air application sended to your email. But I'm not sure that you can reproduce it without "previous versions usage" or some "usage history" because it seems to work fine at "new clear" devices.

Via "grep" only one line could be found in log:
`Jul 19 19:11:42 MacBook-Pro com.apple.xpc.launchd[1] (application.com.adobe.air.ApplicationInstaller.4409402.4409408[6824]) <Warning>: Service exited with abnormal code: 7`

We have a few devices in our hands where it can be reproduced 100%. We can help with tests.

**Comment by ajwfrost** [comment]
Thanks @itlancer  .. we have just tried this .air file here and am getting an exit code 7; internal debugging is showing a problem with the icon sizes when it's trying to generate the .icns file. Not sure why this works on some devices and not others .. will get an update done though so that you can see whether this is the same issue that you're seeing there...

**Comment by ajwfrost** [comment]
Can you please try using the below version of the AIR shared runtime?
https://transfer.harman.com/message/crPP7NlVBgLHPp9LkshSzR

thanks

**Comment by itlancer** [comment]
@ajwfrost 

I figure out that this issue caused by new 256 * 256 icon in application manifest:
`<image256x256>icons/256.png</image256x256>`
Seems something goes wrong this shared runtime for this feature: https://github.com/airsdk/Adobe-Runtime-Support/issues/462
Strange that I can't even install older application versions again after that for some macOS devices.
Building new application version without 256 * 256 icon seems resolve issue with currently latest released AIR Runtime 33.1.1.533.

With AIR Runtime that you sent for some macOS devices event with 256 * 256 icon installation start to works fine. So original issue seems to be resolved by this runtime.
But with some macOS devices (application version with/without 256 * 256 icon and AIR Runtime versions doesn't matter) we got error on application launch (after success installation) "You do not have permission to open the application"
![image](https://user-images.githubusercontent.com/10899066/126336100-58155c51-d3f8-4208-914b-78ec0db87e86.png)
Seems this is issue: https://github.com/airsdk/Adobe-Runtime-Support/issues/565
Quarantine and other changes didn't help for us.

I think this issue should be closed after new AIR runtime version will be released. And waiting for a fix https://github.com/airsdk/Adobe-Runtime-Support/issues/565

**Comment by ajwfrost** [comment]
Hi

Thanks - so, the above issue showing the "you do not have permission" is going to be a problem now that security is tightened up (on a recent Big Sur update by the looks of it). If you code-sign the installed application yourself then it should start working - we just checked that on one of our own tests.

The problem of course is that you can only code-sign your own apps: we couldn't code-sign your app because the bundle ID doesn't match. And it's almost useless to have this limitation for .air applications that you're trying to distribute to multitudes of people....

Which means that the current "installation of AIR packages" process needs to change. Currently the binary file is based on a template that's included in the shared runtime and that just gets copied/renamed to act as the 'launcher' for the new application: this is the file that would need to be signed, but we'll also have to check how the signature process works in case we need to work out a lot more details ahead of installation time... 

TBH the best option is probably to create the applications using the "bundle" target, and code-sign as part of this creation e.g.
```
adt -package [whatever your normal signing commands are] -target bundle [Dev ID signing, see below] MyApplication application.xml MyApp.swf etc...
```
This creates a MyApplication.app file that can be packaged up via pkgbuild or similar for distribution. I'm not 100% sure whether we've got the native codesigning also yet enabled for the `-target native` option which would provide a .dmg file for distribution, I can check that.

The 'native signing' options are then e.g.
```
-storetype keystore -alias "Developer ID Application"
```
which may be enough, or you may need to add into that string to include more of the details. But this means that the output package will be code-signed with your Developer ID Application certificate which means when it's put onto a new computer, it will get trusted (as long as the user has enabled this and doesn't still have the "App Store only" option set!)

**Comment by itlancer** [comment]
Original issue with icons fixed with latest AIR Runtime 33.1.1.743.
But problem with "You do not have permission to open the application" still exists (seems that one https://github.com/airsdk/Adobe-Runtime-Support/issues/565).
We have moved to "bundle" target so it's not a problem for us but may be some other developers suffering about this issue.
